### PR TITLE
destructive operations

### PR DESCRIPTION
**Pull request by patduin**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
   <parent>
     <groupId>com.hotels</groupId>
     <artifactId>hotels-oss-parent</artifactId>
-    <version>2.3.3</version>
```
Comment (patduin) [inline pom.xml]: This is this branch: https://github.com/HotelsDotCom/hotels-oss-parent/pull/23

File: README.md [hunk 2]
```diff
@@ -159,6 +159,7 @@ The table below describes all the available configuration values for Circus Trai
 |`table-replications[n].replica-table.table-name`|No|The name of the table at the destination. Defaults to source table name.|
 |`table-replications[n].copier-options`|No|Table specific `Copier` options which override any global options. See [Copier options](#copier-options) for details.|
 |`table-replications[n].replication-mode`|No|Table replication mode. See [Replication Mode](#replication-mode) for more information. Defaults to `FULL`.|
+|`table-replications[n].replication-strategy`|No|Table replication strategu. See [Replication Strategy](#replication-strategy) for more information. Defaults to `UPSERT.|
```
Comment (patduin) [inline README.md]: strategu???

File: circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/conf/ReplicationStrategy.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.api.conf;
+
+public enum ReplicationStrategy {
+
+  UPSERT,
+  DESTRUCTIVE;
```
Comment (patduin) [inline circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/conf/ReplicationStrategy.java]: PROPAGATE_DELETES

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    this.upsertReplicationFactory = upsertReplicationFactory;
+    this.sourceMetaStoreClientSupplier = sourceMetaStoreClientSupplier;
+    this.replicaMetaStoreClientSupplier = replicaMetaStoreClientSupplier;
+    this.housekeepingListener = housekeepingListener;
+    this.replicaCatalogListener = replicaCatalogListener;
+  }
+
+  @Override
+  public Replication newInstance(TableReplication tableReplication) {
+    String eventId = eventIdFactory.newEventId("ctp");
```
Comment (patduin) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java]: move inside the if, Check where this is used, might be in a lot of places.

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    if (tableReplication.getReplicationStrategy() == ReplicationStrategy.DESTRUCTIVE) {
+      return new DestructiveReplication(upsertReplicationFactory, tableReplication, eventId,
+          new DestructiveSource(sourceMetaStoreClientSupplier, tableReplication), new DestructiveReplica(
+              replicaMetaStoreClientSupplier, getCleanUpLocationManager(eventId, tableReplication), tableReplication));
+    }
+    return upsertReplicationFactory.newInstance(tableReplication);
+  }
+
+  private CleanupLocationManager getCleanUpLocationManager(String eventId, TableReplication tableReplication) {
```
Comment (patduin) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java]: rename method: createDestructiveCleanupLocationManager.

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    if (tableReplication.getReplicationMode() == ReplicationMode.FULL) {
+      return new HouseKeepingCleanupLocationManager(eventId, housekeepingListener, replicaCatalogListener,
+          tableReplication.getReplicaDatabaseName(), tableReplication.getReplicaTableName());
+    } else {
+      return CleanupLocationManager.NULL_CLEANUP_LOCATION_MANAGER;
```
Comment (patduin) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java]: add a comment: Metadata Mirror and Metadata Update should not delete files, so dummy CleanupLocationManager is created.

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplica.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+      CleanupLocationManager cleanupLocationManager,
+      TableReplication tableReplication) {
+    this.replicaMetaStoreClientSupplier = replicaMetaStoreClientSupplier;
+    this.cleanupLocationManager = cleanupLocationManager;
+    this.tableReplication = tableReplication;
+    databaseName = tableReplication.getReplicaDatabaseName();
+    tableName = tableReplication.getReplicaTableName();
+  }
+
+  public boolean tableExists() throws TException {
```
Comment (patduin) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplica.java]: Check if this is needed

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManager.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      String replicaTable) {
+    this.eventId = eventId;
+    this.housekeepingListener = housekeepingListener;
+    this.replicaCatalogListener = replicaCatalogListener;
+    this.replicaDatabase = replicaDatabase;
+    this.replicaTable = replicaTable;
+  }
+
+  @Override
+  public void cleanUpLocations() throws CircusTrainException {
```
Comment (patduin) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManager.java]: cleanUp -> cleanup

Comment (patduin) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManager.java]: rename to: registerLocations

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManager.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        uris.add(location.path.toUri());
+      }
+      replicaCatalogListener.deprecatedReplicaLocations(uris);
+    } finally {
+      locations.clear();
+    }
+  }
+
+  @Override
+  public void addCleanUpLocation(String pathEventId, Path location) {
```
Comment (patduin) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManager.java]: -> addLocation

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManager.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    LOG.debug("Adding clean up location: {}", location.toUri());
+    locations.add(new CleanUpLocation(pathEventId, location, replicaDatabase, replicaTable));
+  }
+
+  private static class CleanUpLocation {
```
Comment (patduin) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManager.java]: CleanupLocation

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplica.java (hunk 7 above)
Comment (patduin) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplica.java]: removed

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    this.sourceMetaStoreClientSupplier = sourceMetaStoreClientSupplier;
+    this.replicaMetaStoreClientSupplier = replicaMetaStoreClientSupplier;
+    this.housekeepingListener = housekeepingListener;
+    this.replicaCatalogListener = replicaCatalogListener;
+  }
+
+  @Override
+  public Replication newInstance(TableReplication tableReplication) {
+    if (tableReplication.getReplicationStrategy() == ReplicationStrategy.PROPAGATE_DELETES) {
+      String eventId = eventIdFactory.newEventId("ctdestructive");
```
Comment (massdosage) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java]: Should we make this `ctd` to at least be consistent with the others which are also three letters?

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManager.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hotels.bdp.circustrain.api.CircusTrainException;
+import com.hotels.bdp.circustrain.api.event.ReplicaCatalogListener;
+import com.hotels.bdp.circustrain.api.listener.HousekeepingListener;
+
+public class HouseKeepingCleanupLocationManager implements CleanupLocationManager {
```
Comment (massdosage) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManager.java]: HouseKeepingCleanupLocationManager -> HousekeepingCleanupLocationManager

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplicaTest.java [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+    table.setTableName(REPLICA_TABLE);
+    table.setPartitionKeys(Lists.newArrayList(new FieldSchema("part1", "string", "")));
+    Map<String, String> parameters = new HashMap<>();
+    parameters.put(CircusTrainTableParameter.SOURCE_TABLE.parameterName(), DATABASE + "." + TABLE);
+    parameters.put(REPLICATION_EVENT.parameterName(), EVENT_ID);
+    table.setParameters(parameters);
+    StorageDescriptor sd1 = new StorageDescriptor();
+    sd1.setLocation(tableLocation.toString());
+    table.setSd(sd1);
+
```
Comment (massdosage) [inline circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplicaTest.java]: There it is!

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManagerTest.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import com.google.common.collect.Lists;
+
+import com.hotels.bdp.circustrain.api.event.ReplicaCatalogListener;
+import com.hotels.bdp.circustrain.api.listener.HousekeepingListener;
+
+@RunWith(MockitoJUnitRunner.class)
+public class HouseKeepingCleanupLocationManagerTest {
```
Comment (massdosage) [inline circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/HouseKeepingCleanupLocationManagerTest.java]: HouseKeepingCleanupLocationManagerTest -> HousekeepingCleanupLocationManagerTest

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java (hunk 11 above)
Comment (patduin) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java]: I'm ok with ctd, sure

File: circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplicaTest.java (hunk 13 above)
Comment (patduin) [reply circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplicaTest.java]: WALLY/WALDO!

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java (hunk 11 above)
Comment (patduin) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/StrategyBasedReplicationFactory.java]: I'm gonna make an enum of CTD/CTP/CTT and have full names so this is all a bit more understandable

File: circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplica.java [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+      }
+      String sourceTableParameterValue = client
+          .getTable(databaseName, tableName)
+          .getParameters()
+          .get(CircusTrainTableParameter.SOURCE_TABLE.parameterName());
+      if (sourceTableParameterValue != null) {
+        String qualifiedName = tableReplication.getSourceTable().getQualifiedName();
+        return qualifiedName.equals(sourceTableParameterValue);
+      }
+
```
Comment (AnanaMJ) [inline circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplica.java]: Can you remove this empty line please?

Comment (patduin) [reply circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/DestructiveReplica.java]: gone

**Pull request merged**